include/boost/compute/algorithm/detail/copy_on_device.hpp

```diff
@@ -27,93 +27,96 @@ namespace boost {
+                                size_t count,
+                                command_queue &queue)
 {
-public:
-    copy_kernel(const device &device)
-        : meta_kernel("copy")
-    {
-        m_count = 0;
-
-        typedef typename std::iterator_traits<InputIterator>::value_type input_type;
-
-        boost::shared_ptr<parameter_cache> parameters =
-            detail::parameter_cache::get_global_cache(device);
+    typedef typename std::iterator_traits<InputIterator>::value_type input_type;
+
+    const device& device = queue.get_device();
+    boost::shared_ptr<parameter_cache> parameters =
+        detail::parameter_cache::get_global_cache(device);
+    std::string cache_key =
+        "__boost_copy_kernel_" + boost::lexical_cast<std::string>(sizeof(input_type));
+
+    uint_ vpt = parameters->get(cache_key, "vpt", 4);
+    uint_ tpb = parameters->get(cache_key, "tpb", 128);
+
+    meta_kernel k("copy");
+    k <<
+        "uint index = get_local_id(0) + " <<
+            "(" << vpt * tpb << " * get_group_id(0));\n" <<
+        "for(uint i = 0; i < " << vpt << "; i++){\n" <<
+        "    if(index < count){\n" <<
+                result[k.var<uint_>("index")] << '=' <<
+                    first[k.var<uint_>("index")] << ";\n" <<
+        "       index += " << tpb << ";\n"
+        "    }\n"
+        "}\n";
+
+    k.add_set_arg<const uint_>("count", static_cast<uint_>(count));
+    size_t global_work_size = calculate_work_size(count, vpt, tpb);
+    return k.exec_1d(queue, 0, global_work_size, tpb);
+}
 
-        std::string cache_key =
-            "__boost_copy_kernel_" + boost::lexical_cast<std::string>(sizeof(input_type));
+template<class InputIterator, class OutputIterator>
+inline event dispatch_copy_on_device(InputIterator first,
+                                     InputIterator last,
+                                     OutputIterator result,
+                                     command_queue &queue)
+{
+    const size_t count = detail::iterator_range_size(first, last);
 
-        m_vpt = parameters->get(cache_key, "vpt", 4);
-        m_tpb = parameters->get(cache_key, "tpb", 128);
+    if(count == 0){
+        // nothing to do
+        return event();
     }
 
-    void set_range(InputIterator first,
-                   InputIterator last,
-                   OutputIterator result)
-    {
-        m_count_arg = add_arg<uint_>("count");
-
-        *this <<
-            "uint index = get_local_id(0) + " <<
-               "(" << m_vpt * m_tpb << " * get_group_id(0));\n" <<
-            "for(uint i = 0; i < " << m_vpt << "; i++){\n" <<
-            "    if(index < count){\n" <<
-                     result[expr<uint_>("index")] << '=' <<
-                         first[expr<uint_>("index")] << ";\n" <<
-            "        index += " << m_tpb << ";\n"
-            "    }\n"
-            "}\n";
-
-        m_count = detail::iterator_range_size(first, last);
+#ifndef __APPLE__
```

Comment:
Hmm, I wonder if it would be better to use run-time dispatching (something like `device.platform().name() == "Apple"`) rather than compile-time dispatching (probably unlikely, but something that is compiled on Apple host may be executed on a non-Apple compute device or vice-versa)?